scipy/special/xsf/tools.h

```diff
@@ -274,10 +271,163 @@ namespace detail {
      * See ContinuedFractionSeriesGenerator for details.
      */
     template <typename Generator, typename T = pair_value_t<generator_result_t<Generator>>>
-    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T>
-    continued_fraction_series(Generator &cf) {
+    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T> continued_fraction_series(Generator &cf) {
         return ContinuedFractionSeriesGenerator<Generator, T>(cf);
     }
 
+    /* Find initial bracket for a bracketing scalar root finder. A valid bracket is a pair of points a < b for
+     * which the signs of f(a) and f(b) differ. If f(x0) = 0, where x0 is the initial guess, this bracket finder
+     * will return the bracket (x0, x0). It is expected that the rootfinder will check if the bracket
+     * endpoints are roots.
+     *
+     * This is a private function intended specifically for the situation where
+     * the goal is to invert a CDF function F for a parametrized family of distributions with respect to one
+     * parameter, when the other parameters are known, and where F is monotonic with respect to the unknown parameter.
+     *
+     * Note that this takes a pointer to a function taking a tuple of args along with a scalar
+     * double argument. A tuple of args for specializing func is also passed as the final argument.
+     * It would be much cleaner to use std::function and capturing lambda's to specialize the
+     * function we are finding a root for, but I couldn't get this working in CuPy using NVRTC.
+     * This should be revisited in the future in order to allow simplifying this code. */
+    template <typename Function>
+    XSF_HOST_DEVICE inline std::tuple<double, double, double, double, int> bracket_root_for_cdf_inversion(
+        Function func, double x0, double xmin, double xmax, double step0_left,
+        double step0_right, double factor_left, double factor_right, bool increasing, std::uint64_t maxiter
+    ) {
+        double y0 = func(x0);
+
+        if (y0 == 0) {
+            // Initial guess is correct.
+            std::tuple<double, double, double, double, int> result(x0, x0, y0, y0, 0);
+            return result;
+        }
```

Comment:
> * Could you give an example illustrating "If f(x0) = 0, where x0 is the initial guess, this bracket finder will return the bracket (x0, x0)."

Here are the lines where the bracket finder checks if the initial guess is correct. For an example where this actually happens, we can consider:

```Python
gdtr(0.125, 1.0, 0.125)
# np.float64(0.0155035629945916)

gdtr(0.125, 0.0155035629945916, 0.125)
# np.float64(1.0)
```

The initial guess for the rootfinder is 1.0 which happens to be correct in this case. A quick spot check is that `gdtrib` is faster here than it is when the initial guess is a little off 

```
In [38]: %timeit gdtrib(0.125, 0.0155035629945916, 0.125)
500 ns ± 0.248 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [39]: %timeit gdtrib(0.125, 0.016, 0.125)
767 ns ± 1.51 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Reply:
Perfect, thanks!